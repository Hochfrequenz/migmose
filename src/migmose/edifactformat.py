"""
Extend the existing edifact format enum from maus.edifact with the formats UTILMDS, UTILMDG
for Strom, Gas respectively.
"""

from maus.edifact import EdifactFormat


class ExtendedEdifactFormat(EdifactFormat):
    UTILMDS = "UTILMDS"  #: utilities master data, Strom
    UTILMDG = "UTILMDG"  #: utilities master data, Gas
