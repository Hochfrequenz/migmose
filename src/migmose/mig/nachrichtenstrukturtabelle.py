"""
contains class mig tables
"""

import csv
from pathlib import Path

from loguru import logger
from maus.edifact import EdifactFormat
from pydantic import BaseModel

from migmose.mig.nachrichtenstrukturzeile import NachrichtenstrukturZeile


class NachrichtenstrukturTabelle(BaseModel):
    """
    class for mig tables
    """

    lines: list[NachrichtenstrukturZeile]

    @classmethod
    def init_raw_table(cls, raw_lines: list[str]) -> "NachrichtenstrukturTabelle":
        """
        reads table as list of raw lines and returns a NachrichtenstrukturTabelle
        consisting of NachrichtenstrukturZeilen
        """
        collected_lines: list[NachrichtenstrukturZeile] = []
        for raw_line in raw_lines:
            collected_lines.append(NachrichtenstrukturZeile.init_raw_lines(raw_line))
        return cls(lines=collected_lines)

    def output_as_csv(self, message_type: EdifactFormat, output_dir: Path) -> None:
        """
        writes the NestedNachrichtenstruktur as json
        """
        output_dir.mkdir(parents=True, exist_ok=True)
        file_path = output_dir.joinpath(f"{message_type}_nachrichtenstruktur.csv")
        fieldnames = list(NachrichtenstrukturZeile.model_json_schema()["properties"].keys())
        with open(file_path, "w", encoding="utf-8") as csv_file:
            writer = csv.DictWriter(csv_file, fieldnames=fieldnames)
            writer.writeheader()
            for nachrichtenstruktur_zeile in self.lines:
                writer.writerow(nachrichtenstruktur_zeile.model_dump())
        logger.info(f"Wrote Nachrichtenstruktur for {message_type} to {file_path}")
