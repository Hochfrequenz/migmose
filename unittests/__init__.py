"""
This file is here, because this allows for best de-coupling of tests and application/library logic.
Further reading: https://docs.pytest.org/en/6.2.x/goodpractices.html#tests-outside-application-code
"""

from pathlib import Path

path_to_test_edi_energy_mirror_repo = Path(__file__).parent / "test_edi_energy_mirror_repo"
path_to_test_FV2310 = path_to_test_edi_energy_mirror_repo / "edi_energy_de" / "FV2310"
expected_output_dir = Path(__file__).parent / "test_expected_data"
