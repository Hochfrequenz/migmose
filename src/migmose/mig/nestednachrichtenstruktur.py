"""
contains class for structured segmentgroups in mig tables. Builds table recursively.
"""

import json
from pathlib import Path
from typing import Optional, Tuple

from loguru import logger
from maus.edifact import EdifactFormat
from pydantic import BaseModel

from migmose.mig.nachrichtenstruktur import NachrichtenstrukturTabelle
from migmose.mig.nachrichtenstrukturzeile import NachrichtenstrukturZeile


class NestedNachrichtenstruktur(BaseModel):
    """
    class for structured segmentgroups in mig tables. Builds table recursively. Inherits from NachrichtenstrukturZeile
    e.g.(ORDCHG):
    {
    "segmente": [
        {
        "zaehler": "0160",
        "nr": "7",
        "bezeichnung": "NAD",
        "standard_status": "M",
        "bdew_status": "M",
        "standard_maximale_wiederholungen": 1,
        "bdew_maximale_wiederholungen": 1,
        "ebene": 1,
        "inhalt": "MP-ID Absender"
        }
        ],
    "segmentgruppen": [
    {
        "segmente": [
        {
            "zaehler": "0260",
            "nr": "8",
            "bezeichnung": "CTA",
            "standard_status": "M",
            "bdew_status": "M",
            "standard_maximale_wiederholungen": 1,
            "bdew_maximale_wiederholungen": 1,
            "ebene": 2,
            "inhalt": "Ansprechpartner"
            },
            {
            "zaehler": "0270",
            "nr": "9",
            "bezeichnung": "COM",
            "standard_status": "C",
            "bdew_status": "R",
            "standard_maximale_wiederholungen": 5,
            "bdew_maximale_wiederholungen": 5,
            "ebene": 3,
            "inhalt": "Kommunikationsverbindung"
            }
            ],
        "segmentgruppen": []
        }
        ]
    }
    """

    header_linie: Optional[NachrichtenstrukturZeile] = None
    segmente: list[Optional[NachrichtenstrukturZeile]] = []
    segmentgruppen: list[Optional["NestedNachrichtenstruktur"]] = []

    @classmethod
    def structure_table(
        cls, table: NachrichtenstrukturTabelle, segmentgruppe: Optional[NachrichtenstrukturZeile] = None, index: int = 0
    ) -> Tuple["NestedNachrichtenstruktur", int]:
        """init structured table"""
        collected_segments: list[Optional[NachrichtenstrukturZeile]] = []
        collected_segmentgroups: list[Optional["NestedNachrichtenstruktur"]] = []
        i = index
        while i < len(table.lines):
            line = table.lines[i]
            is_line_segmentgruppe = line.nr is None
            if is_line_segmentgruppe:
                added_segmentgroup, i = cls.structure_table(table, line, i + 1)
                collected_segmentgroups.append(added_segmentgroup)
            else:
                collected_segments.append(line)
                i += 1
            if i < len(table.lines):
                next_line_segmentgruppe = table.lines[i].nr is None

                if segmentgruppe is not None and (
                    (next_line_segmentgruppe and table.lines[i].ebene <= segmentgruppe.ebene)  # type: ignore [operator]
                    or (line.ebene > table.lines[i].ebene)  # type: ignore [operator]
                ):
                    return (
                        cls(
                            header_linie=segmentgruppe,
                            segmente=collected_segments,
                            segmentgruppen=collected_segmentgroups,
                        ),
                        i,
                    )
        return cls(header_linie=segmentgruppe, segmente=collected_segments, segmentgruppen=collected_segmentgroups), i

    @classmethod
    def output_as_json(
        cls, instance: "NestedNachrichtenstruktur", message_type: EdifactFormat, output_dir: Path
    ) -> None:
        """
        writes the NestedNachrichtenstruktur as json
        """

        if not output_dir.exists():
            output_dir.mkdir(parents=True, exist_ok=True)
        file_path = output_dir.joinpath(f"{message_type}_Nested_nachrichtenstruktur.json")
        structured_json = instance.model_dump()
        with open(file_path, "w", encoding="utf-8") as json_file:
            json.dump(structured_json, json_file, indent=4)
        logger.info(f"Wrote nested Nachrichtenstruktur for {message_type} to {file_path}")
