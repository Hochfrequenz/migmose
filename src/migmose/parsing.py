"""
contains functions for file handling and parsing.
"""

import re
from collections import defaultdict
from datetime import datetime
from pathlib import Path
from typing import DefaultDict, Generator, Union

import click
import docx
from docx.document import Document
from docx.oxml import CT_Tbl  # type:ignore[attr-defined]
from docx.table import Table, _Cell
from efoli import EdifactFormat, EdifactFormatVersion
from loguru import logger

from migmose.mig.segmentlayout import SegmentLayout, SegmentLayoutLine


def find_file_to_format(
    message_formats: list[EdifactFormat], edi_energy_repo: Path, format_version: EdifactFormatVersion
) -> dict[EdifactFormat, Path]:
    """
    finds the file with the message type in the input directory
    """
    input_dir = edi_energy_repo / Path("edi_energy_de") / Path(format_version)
    all_file_dict: DefaultDict[EdifactFormat, list[Path]] = defaultdict(list)
    file_dict: dict[EdifactFormat, Path] = {}
    for message_format in message_formats:
        for file in input_dir.iterdir():
            if "MIG" not in file.name or file.suffix != ".docx":
                continue
            if message_format is EdifactFormat.UTILMDG and "Gas" in file.name:
                all_file_dict[EdifactFormat.UTILMDG].append(file)
            elif message_format is EdifactFormat.UTILMDS and "Strom" in file.name:
                all_file_dict[EdifactFormat.UTILMDS].append(file)
            elif message_format in file.name:
                all_file_dict[message_format].append(file)
        if len(all_file_dict[message_format]) == 0:
            logger.warning(f"⚠️ No file found for {message_format}", fg="red")
            continue
        file_dict[message_format] = get_latest_file(all_file_dict[message_format])
    if file_dict:
        return file_dict
    logger.error("❌ No files found in the input directory.", fg="red")
    raise click.Abort()


def _extract_date(file_path: Path) -> tuple[datetime, Path]:
    # Regex to extract the date format YYYYMMDD from the filename as a string
    match = re.search(r"(\d{8})\.docx$", file_path.name)
    if match:
        # Return the date as a datetime object for comparison and the path for use
        return datetime.strptime(match.group(1), "%Y%m%d"), file_path
    logger.warning(
        f"⚠️ No timestamp in filename found in {file_path}."
        + "in case of multiple docx files in this path, it must be a "
        + "timestamp with format 'yyyyMMdd.docx' in filename.",
        fg="red",
    )
    raise click.Abort()


def get_latest_file(file_list: list[Path]) -> Path:
    """
    This function takes a list of docx files Path
    and returns the Path of the latest MIG docx file based on the timestamp in its name.
    The timestamp is assumed to be formatted as YYYYMMDD and located just before the ".docx" extension.

    Parameters:
        file_list (list of Path): A list containing file paths with timestamps.

    Returns:
        Path: The path of the latest file. Returns None if no valid date is found.
    """
    # Initialize variables to keep track of the latest file and date
    latest_file: Path
    latest_date: datetime | None = None

    for file_path in file_list:
        date, path = _extract_date(file_path)
        if latest_date is None or date > latest_date:
            latest_file = path
            latest_date = date

    # Return the path of the file with the latest date
    return latest_file


def get_paragraphs_up_to_diagram(parent: Union[Document, _Cell]) -> Generator[Table, None, None]:
    """Goes through paragraphs and tables"""
    # pylint: disable=protected-access
    if isinstance(parent, Document):
        parent_elm = parent.element.body
    elif isinstance(parent, _Cell):
        parent_elm = parent._tc
    else:
        raise ValueError("Passed parent argument must be of type Document or _Cell")

    for child in parent_elm.iterchildren():
        if isinstance(child, CT_Tbl):
            yield Table(child, parent)


_row_regex = re.compile(r"^(?P<left>\t\d+\t)(?P<nr>\d{0,5})(?P<right>\t.*)$")
"""
https://regex101.com/r/vtF07B/2
"""


def _zfill_nr(row_str: str) -> str:
    match = _row_regex.match(row_str)
    if not match:
        return row_str
    left = match.group("left")
    nr = match.group("nr")
    right = match.group("right")
    return f"{left}{nr.zfill(5)}{right}"


def parse_raw_nachrichtenstrukturzeile(input_path: Path) -> tuple[list[str], dict[str, list[SegmentLayout]]]:
    """
    parses raw nachrichtenstrukturzeile from a table. returns list of raw lines
    """
    # pylint: disable=protected-access
    doc = docx.Document(str(input_path.absolute()))
    docx_objects = get_paragraphs_up_to_diagram(doc)
    mig_tables = []
    nachrichtenstruktur_header = "Status\tMaxWdh\n\tZähler\tNr\tBez\tSta\tBDEW\tSta\tBDEW\tEbene\tInhalt"
    segmentlayout_header = "\tStandard\tBDEW\n\tZähler\tNr\tBez\tSt\tMaxWdh\tSt\tMaxWdh\tEbene\tName"
    segmentlayout_tables = []
    for docx_object in docx_objects:
        segmentlayout_table = []
        for ind, line in enumerate(docx_object._cells):
            # marks the beginning of the complete nachrichtentruktur table
            if line.text == nachrichtenstruktur_header:
                mig_tables.extend([row.text for row in docx_object._cells[ind + 1 :]])
                break

        is_segmentlayout_table = docx_object._cells[0].text == segmentlayout_header
        if is_segmentlayout_table:
            for row in docx_object.rows:
                for cell in iter_visual_cells(row):
                    segmentlayout_table.append(cell)
        if segmentlayout_table:
            segmentlayout_tables.append(segmentlayout_table)
    # filter empty rows and headers
    mig_tables = [_zfill_nr(row) for row in mig_tables if row not in ("", "\n", nachrichtenstruktur_header)]
    segmentlayouts = process_segmentlayouts(segmentlayout_tables)
    return mig_tables, segmentlayouts


def iter_visual_cells(row: Table) -> Generator[_Cell, None, None]:
    """
    Iterate over visual cells in one row. Taken from https://github.com/python-openxml/python-docx/issues/344.
    """
    prior_tc = None
    for cell in row.cells:  # type:ignore [attr-defined]
        this_tc = cell._tc  # pylint: disable=protected-access
        if this_tc is prior_tc:  # skip cells pointing to same `<w:tc>` element
            continue
        yield cell
        prior_tc = this_tc


# pylint: disable=too-many-locals
def process_segmentlayouts(segmentlayout_tables: list[list[_Cell]]) -> dict[str, list[SegmentLayout]]:
    """
    Create Segmentlayouts from list of _Cell objects
    """
    segment_layouts_dict: defaultdict[str, list[SegmentLayout]] = defaultdict(list)
    for segmentlayout_table in segmentlayout_tables:
        start_index_annotations = 0
        start_index_example = 0
        ignore_cells = []
        for index, cell in enumerate(segmentlayout_table):
            cell_text = cell.text
            if cell_text == "Bez":
                ignore_cells.append(index)
            if cell_text == "Bemerkung:":
                ignore_cells.append(index)
                start_index_annotations = index
            if cell_text == "Beispiel:":
                start_index_example = index + 1
                break
        assert len(ignore_cells) != 0

        segment_zaehler = segmentlayout_table[ignore_cells[0] - 6].text.split("\t")[1]

        cell_index = []
        for index, header_ind in enumerate(ignore_cells[:-1]):
            cell_index.extend(list(range(header_ind + 7, ignore_cells[index + 1] - 6, 7)))
        unique_indentations = {
            segmentlayout_table[index].paragraphs[0].paragraph_format.left_indent
            for index in cell_index
            if segmentlayout_table[index].text != ""
        }
        indentations = {indentation: index for index, indentation in enumerate(sorted(unique_indentations))}
        raw_lines = []
        for i in cell_index:
            if segmentlayout_table[i].text != "":
                raw_lines.append(
                    SegmentLayoutLine(
                        bezeichnung=segmentlayout_table[i].text,
                        name=segmentlayout_table[i + 1].text,
                        standard_status=segmentlayout_table[i + 2].text,
                        standard_format=segmentlayout_table[i + 3].text,
                        bdew_status=segmentlayout_table[i + 4].text.replace("\t", ""),
                        bdew_format=segmentlayout_table[i + 5].text.replace("\t", ""),
                        anwendung=segmentlayout_table[i + 6].text,
                        indent=indentations[segmentlayout_table[i].paragraphs[0].paragraph_format.left_indent],
                    )
                )
            else:
                raw_lines[-1].bezeichnung += segmentlayout_table[i].text
                raw_lines[-1].name += segmentlayout_table[i + 1].text
                raw_lines[-1].standard_status += segmentlayout_table[i + 2].text
                raw_lines[-1].bdew_status += segmentlayout_table[i + 3].text
                raw_lines[-1].standard_format += segmentlayout_table[i + 4].text
                raw_lines[-1].bdew_format += segmentlayout_table[i + 5].text
                raw_lines[-1].anwendung += segmentlayout_table[i + 6].text
        segment_layouts_dict[segment_zaehler].append(
            SegmentLayout(
                struktur=raw_lines,
                bemerkung="".join(
                    cell.text for cell in segmentlayout_table[start_index_annotations : start_index_example - 1]
                ),
                beispiel="".join(cell.text for cell in segmentlayout_table[start_index_example:]),
            )
        )
    return segment_layouts_dict
