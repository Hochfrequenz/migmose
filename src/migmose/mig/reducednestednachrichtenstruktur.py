"""
contains class for trees consisting of segments of mig tables
"""

import json
from pathlib import Path
from typing import Any, Optional

from loguru import logger
from maus.edifact import EdifactFormat
from pydantic import BaseModel

from migmose.mig.nachrichtenstrukturzeile import NachrichtenstrukturZeile
from migmose.mig.nestednachrichtenstruktur import NestedNachrichtenstruktur


class ReducedNestedNachrichtenstruktur(BaseModel):
    """will contain the tree structure of nachrichtenstruktur tables"""

    header_linie: Optional[NachrichtenstrukturZeile] = None
    segmente: list[Optional[NachrichtenstrukturZeile]] = []
    segmentgruppen: list[Optional["ReducedNestedNachrichtenstruktur"]] = []

    @classmethod
    def create_reduced_nested_nachrichtenstruktur(
        cls, nachrichten_struktur: NestedNachrichtenstruktur
    ) -> "ReducedNestedNachrichtenstruktur":
        """init nested Nachrichtenstruktur"""

        # Helper function to create a unique identifier for each segment
        def get_identifier(segment: Optional[NachrichtenstrukturZeile]) -> tuple[str, str]:
            if segment is None:
                return ("0", "root")
            return (segment.zaehler, segment.bezeichnung)

        def count_segments(segment_group: ReducedNestedNachrichtenstruktur | NestedNachrichtenstruktur) -> int:
            # Start with counting segments directly under the current segment group
            total_segments = len(segment_group.segmente)
            # Recursively count segments in nested segment groups
            for nested_sg in segment_group.segmentgruppen:
                if nested_sg is not None:
                    total_segments += count_segments(nested_sg)

            return total_segments

        # Function to process segments and remove duplicates within the same list.
        def process_segments(
            segments: list[Optional[NachrichtenstrukturZeile]],
        ) -> list[Optional[RducedNachrichtenstrukturZeile]]:
            seen = set()
            unique_segments = []
            for segment in segments:
                if segment is not None:
                    identifier = get_identifier(segment)
                    if identifier not in seen:
                        seen.add(identifier)
                        unique_segments.append(segment)
            return unique_segments

        # Recursive function to traverse and clean segment groups
        def process_segmentgruppen(
            segmentgruppen: list[Optional[NestedNachrichtenstruktur]],
            segment_count_dict: dict,
            seen=None,
            depth: int = 0,
        ):
            """Recursively clean segment groups to avoid duplicates, keep largest,
            with debugging for circular references."""
            if seen is None:
                seen = {}
            result = []

            for sg in segmentgruppen:
                if sg is not None:
                    identifier = get_identifier(sg.header_linie)
                    max_count, max_sg = segment_count_dict[identifier]

                    if identifier not in seen:
                        seen[identifier] = max_sg
                        logger.info("Added {} with {} segments at depth {}.", identifier, max_count, depth)

                    sg.segmente = process_segments(max_sg.segmente)
                    sg.segmentgruppen = process_segmentgruppen(
                        max_sg.segmentgruppen, segment_count_dict, seen, depth + 1
                    )

            # Compile the unique list from the seen dictionary after recursive processing to avoid circular reference
            if depth == 0:  # Only compile on the initial call, not recursive ones
                result = [seen[key] for key in seen]
            return result

        def build_segment_count_dict(segment_groups: list[Optional[NestedNachrichtenstruktur]]):
            segment_count_dict: dict = {}
            for sg in segment_groups:
                if sg is not None:
                    name = get_identifier(sg.header_linie)
                    count = count_segments(sg)

                    # Check if the current segment group's count is greater than the stored count
                    if name in segment_count_dict:
                        existing_count, _ = segment_count_dict[name]
                        if count > existing_count:
                            segment_count_dict[name] = (count, sg)
                    else:
                        segment_count_dict[name] = (count, sg)

                    # Process nested segment groups recursively and update the dictionary
                    nested_counts = build_segment_count_dict(sg.segmentgruppen)
                    for nested_name, (nested_count, nested_sg) in nested_counts.items():
                        if nested_name in segment_count_dict:
                            existing_count, _ = segment_count_dict[nested_name]
                            if nested_count > existing_count:
                                segment_count_dict[nested_name] = (nested_count, nested_sg)
                        else:
                            segment_count_dict[nested_name] = (nested_count, nested_sg)

            return segment_count_dict

        data = ReducedNestedNachrichtenstruktur()
        # Start processing the top-level segments
        if nachrichten_struktur.segmente is not None:
            data.segmente = process_segments(nachrichten_struktur.segmente)

        # Process segment groups recursively
        if nachrichten_struktur.segmentgruppen is not None:
            segment_count_dict = build_segment_count_dict(nachrichten_struktur.segmentgruppen)
            data.segmentgruppen = process_segmentgruppen(nachrichten_struktur.segmentgruppen, segment_count_dict)

        return data

    def to_json(self, message_type: EdifactFormat, output_dir: Path) -> dict[str, Any]:
        """
        writes the reduced NestedNachrichtenstruktur as json
        """
        output_dir.mkdir(parents=True, exist_ok=True)
        file_path = output_dir.joinpath(f"{message_type}_reduced_nested_nachrichtenstruktur.json")
        structured_json = self.model_dump()
        with open(file_path, "w", encoding="utf-8") as json_file:
            json.dump(structured_json, json_file, indent=4)
        logger.info("Wrote reduced nested Nachrichtenstruktur for {} to {}", message_type, file_path)
        return structured_json
