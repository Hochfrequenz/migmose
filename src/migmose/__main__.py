"""
contains CLI logic for migmose.
"""

from pathlib import Path

import click
from loguru import logger
from maus.edifact import EdifactFormat

from migmose.mig.nachrichtenstrukturtabelle import NachrichtenstrukturTabelle
from migmose.mig.nestednachrichtenstruktur import NestedNachrichtenstruktur
from migmose.parsing import find_file_to_format, parse_raw_nachrichtenstrukturzeile


# add CLI logic
@click.command()
@click.option(
    "-i",
    "--input-dir",
    type=click.Path(exists=True, dir_okay=True, file_okay=False, path_type=Path),
    prompt="Please enter the path to the directory containing the .docx files",
    help="Set path to directory which contains the .docx files for the migs",
)
@click.option(
    "-mf",
    "--message-format",
    type=click.Choice(list(map(lambda x: x.name, EdifactFormat)), case_sensitive=False),
    # Taken from https://github.com/pallets/click/issues/605#issuecomment-889462570
    default=list(map(lambda x: x.name, EdifactFormat)),
    help="Defines the set of message formats to be parsed.",
    multiple=True,
)
@click.option(
    "-o",
    "--output-dir",
    type=click.Path(exists=False, dir_okay=True, file_okay=False, path_type=Path),
    prompt="Please enter the path to the directory which should contain the output files.",
    help="Set path to directory which contains the output files. If the directory does not exist, it will be created.",
)
@click.option(
    "-ft",
    "--file-type",
    type=click.Choice(["csv", "nested_json"], case_sensitive=False),
    default=["csv"],
    prompt="Please specify how the output should be formatted.",
    help="Defines the output format. Choose between csv and nested_json. Default is csv.",
    multiple=True,
)
def main(input_dir: Path, output_dir, message_format: list[EdifactFormat], file_type: list[str]) -> None:
    """
    Main function. Uses CLI input.
    """
    if message_format is None:
        message = "❌ No message format specified. Please specify the message format."
        click.secho(message, fg="yellow")
        logger.error(message)
        raise click.Abort()
    if file_type is None:
        message = "❌ No output format specified. Please specify the output format."
        click.secho(message, fg="yellow")
        logger.error(message)
        raise click.Abort()

    dict_files = find_file_to_format(message_format, input_dir)
    for m_format, file in dict_files.items():
        raw_lines = parse_raw_nachrichtenstrukturzeile(file)
        nachrichtenstrukturtabelle = NachrichtenstrukturTabelle.create_nachrichtenstruktur_tabelle(raw_lines)
        if "csv" in file_type:
            logger.info("💾 Saving flat Nachrichtenstruktur table for {} as csv to {}.", m_format, output_dir)
            nachrichtenstrukturtabelle.to_csv(m_format, output_dir)
        nested_nachrichtenstruktur, _ = NestedNachrichtenstruktur.create_nested_nachrichtenstruktur(
            nachrichtenstrukturtabelle
        )
        if "nested_json" in file_type:
            logger.info("💾 Saving nested Nachrichtenstruktur for {} as json to {}.", m_format, output_dir)
            nested_nachrichtenstruktur.to_json(m_format, output_dir)


if __name__ == "__main__":
    main()  # pylint:disable=no-value-for-parameter
