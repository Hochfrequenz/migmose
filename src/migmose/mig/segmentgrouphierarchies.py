"""
contains class for trees consisting of segments of mig tables
"""

import json
from pathlib import Path
from typing import Any, Optional

from loguru import logger
from maus.edifact import EdifactFormat
from pydantic import BaseModel, Field

from migmose.mig.reducednestednachrichtenstruktur import ReducedNestedNachrichtenstruktur


class SegmentGroupHierarchy(BaseModel):
    """Contains the model for a segment group hierarchy used by the MAUS library."""

    opening_segment: Optional[str] = None
    segment_group: Optional[str] = None
    sub_hierarchy: list[Optional["SegmentGroupHierarchy"]] = Field(default_factory=list)

    @classmethod
    def create_segmentgroup_hierarchy(
        cls, reduced_nested_nachrichtenstruktur: ReducedNestedNachrichtenstruktur
    ) -> "SegmentGroupHierarchy":
        """init Segmentrgroup Hierarchy"""
        segment_group: Optional[str] = None
        opening_segment: Optional[str] = None
        if reduced_nested_nachrichtenstruktur.header_linie is not None:
            segment_group = reduced_nested_nachrichtenstruktur.header_linie.bezeichnung
        if reduced_nested_nachrichtenstruktur.segmente and reduced_nested_nachrichtenstruktur.segmente[0] is not None:
            opening_segment = reduced_nested_nachrichtenstruktur.segmente[0].bezeichnung
        sub_hierarchy: list[Optional["SegmentGroupHierarchy"]] = []
        for sub_segmentgroup in reduced_nested_nachrichtenstruktur.segmentgruppen:
            if sub_segmentgroup is not None:
                sub_hierarchy.append(cls.create_segmentgroup_hierarchy(sub_segmentgroup))

        return SegmentGroupHierarchy(
            segment_group=segment_group, opening_segment=opening_segment, sub_hierarchy=sub_hierarchy
        )

    def to_json(self, message_type: EdifactFormat, output_dir: Path) -> dict[str, Any]:
        """
        writes the reduced NestedNachrichtenstruktur as json
        """
        output_dir.mkdir(parents=True, exist_ok=True)
        file_path = output_dir.joinpath("sgh.json")
        structured_json = self.model_dump()
        with open(file_path, "w", encoding="utf-8") as json_file:
            json.dump(structured_json, json_file, indent=4)
        logger.info("Wrote segmentgroup hierarchy (sgh.json) for {} to {}", message_type, file_path)
        return structured_json
