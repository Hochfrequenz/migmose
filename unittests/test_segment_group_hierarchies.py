import json

import pytest
from efoli import EdifactFormat, EdifactFormatVersion
from maus.models.message_implementation_guide import SegmentGroupHierarchy as MausSGH
from maus.models.message_implementation_guide import SegmentGroupHierarchySchema

from migmose.mig.nachrichtenstrukturtabelle import NachrichtenstrukturTabelle
from migmose.mig.nestednachrichtenstruktur import NestedNachrichtenstruktur
from migmose.mig.reducednestednachrichtenstruktur import ReducedNestedNachrichtenstruktur
from migmose.mig.segmentgrouphierarchies import SegmentGroupHierarchy
from migmose.parsing import find_file_to_format, parse_raw_nachrichtenstrukturzeile
from unittests import path_to_test_edi_energy_mirror_repo


class TestSegmentGroupHierarchy:
    """test class for create_reduced_nested_nachrichtenstruktur"""

    @pytest.mark.parametrize(
        "message_format",
        [
            pytest.param(EdifactFormat.ORDCHG, id="ORDCHG"),
            pytest.param(EdifactFormat.UTILMDS, id="UTILMDS"),
            pytest.param(EdifactFormat.IFTSTA, id="IFTSTA"),
        ],
    )
    def test_create_create_segmentgroup_hierarchy(self, message_format: EdifactFormat, tmp_path, snapshot):
        """test if the reduced nested nachrichtenstruktur is created correctly"""
        file_path_dict = find_file_to_format(
            [message_format], path_to_test_edi_energy_mirror_repo, EdifactFormatVersion.FV2310
        )
        file_path = file_path_dict[message_format]
        (raw_lines, _) = parse_raw_nachrichtenstrukturzeile(file_path)
        nachrichtenstrukturtabelle = NachrichtenstrukturTabelle.create_nachrichtenstruktur_tabelle(raw_lines)
        nested_nachrichtenstruktur, _ = NestedNachrichtenstruktur.create_nested_nachrichtenstruktur(
            nachrichtenstrukturtabelle
        )
        reduced_nested_nachrichtenstruktur = ReducedNestedNachrichtenstruktur.create_reduced_nested_nachrichtenstruktur(
            nested_nachrichtenstruktur
        )
        sgh = SegmentGroupHierarchy.create_segmentgroup_hierarchy(reduced_nested_nachrichtenstruktur)
        sgh.to_json(message_format, tmp_path)
        with open(tmp_path / f"{message_format}.sgh.json", "r", encoding="utf-8") as file1:
            actual_sgh_json = json.load(file1)
            assert actual_sgh_json == snapshot
        with open(tmp_path / f"{message_format}.sgh.json", "r", encoding="utf-8") as sgh_file:
            sgh = SegmentGroupHierarchySchema().loads(sgh_file.read())
            assert isinstance(sgh, MausSGH)
