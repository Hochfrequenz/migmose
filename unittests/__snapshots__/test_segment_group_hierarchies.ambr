# serializer version: 1
# name: TestSegmentGroupHierarchy.test_create_create_segmentgroup_hierarchy[IFTSTA]
  dict({
    'opening_segment': 'UNH',
    'segment_group': None,
    'sub_hierarchy': list([
      dict({
        'opening_segment': 'NAD',
        'segment_group': 'SG1',
        'sub_hierarchy': list([
          dict({
            'opening_segment': 'CTA',
            'segment_group': 'SG2',
            'sub_hierarchy': list([
            ]),
          }),
        ]),
      }),
      dict({
        'opening_segment': 'EQD',
        'segment_group': 'SG4',
        'sub_hierarchy': list([
          dict({
            'opening_segment': 'LOC',
            'segment_group': 'SG6',
            'sub_hierarchy': list([
            ]),
          }),
          dict({
            'opening_segment': 'STS',
            'segment_group': 'SG7',
            'sub_hierarchy': list([
            ]),
          }),
        ]),
      }),
      dict({
        'opening_segment': 'CNI',
        'segment_group': 'SG14',
        'sub_hierarchy': list([
          dict({
            'opening_segment': 'STS',
            'segment_group': 'SG15',
            'sub_hierarchy': list([
              dict({
                'opening_segment': 'EFI',
                'segment_group': 'SG16',
                'sub_hierarchy': list([
                ]),
              }),
              dict({
                'opening_segment': 'NAD',
                'segment_group': 'SG17',
                'sub_hierarchy': list([
                  dict({
                    'opening_segment': 'CTA',
                    'segment_group': 'SG18',
                    'sub_hierarchy': list([
                    ]),
                  }),
                ]),
              }),
              dict({
                'opening_segment': 'GID',
                'segment_group': 'SG25',
                'sub_hierarchy': list([
                ]),
              }),
            ]),
          }),
        ]),
      }),
    ]),
  })
# ---
# name: TestSegmentGroupHierarchy.test_create_create_segmentgroup_hierarchy[ORDCHG]
  dict({
    'opening_segment': 'UNH',
    'segment_group': None,
    'sub_hierarchy': list([
      dict({
        'opening_segment': 'RFF',
        'segment_group': 'SG1',
        'sub_hierarchy': list([
        ]),
      }),
      dict({
        'opening_segment': 'NAD',
        'segment_group': 'SG3',
        'sub_hierarchy': list([
          dict({
            'opening_segment': 'CTA',
            'segment_group': 'SG6',
            'sub_hierarchy': list([
            ]),
          }),
        ]),
      }),
    ]),
  })
# ---
# name: TestSegmentGroupHierarchy.test_create_create_segmentgroup_hierarchy[UTILMDS]
  dict({
    'opening_segment': 'UNH',
    'segment_group': None,
    'sub_hierarchy': list([
      dict({
        'opening_segment': 'RFF',
        'segment_group': 'SG1',
        'sub_hierarchy': list([
        ]),
      }),
      dict({
        'opening_segment': 'NAD',
        'segment_group': 'SG2',
        'sub_hierarchy': list([
          dict({
            'opening_segment': 'CTA',
            'segment_group': 'SG3',
            'sub_hierarchy': list([
            ]),
          }),
        ]),
      }),
      dict({
        'opening_segment': 'IDE',
        'segment_group': 'SG4',
        'sub_hierarchy': list([
          dict({
            'opening_segment': 'LOC',
            'segment_group': 'SG5',
            'sub_hierarchy': list([
            ]),
          }),
          dict({
            'opening_segment': 'RFF',
            'segment_group': 'SG6',
            'sub_hierarchy': list([
            ]),
          }),
          dict({
            'opening_segment': 'SEQ',
            'segment_group': 'SG8',
            'sub_hierarchy': list([
              dict({
                'opening_segment': 'QTY',
                'segment_group': 'SG9',
                'sub_hierarchy': list([
                ]),
              }),
              dict({
                'opening_segment': 'CCI',
                'segment_group': 'SG10',
                'sub_hierarchy': list([
                ]),
              }),
            ]),
          }),
          dict({
            'opening_segment': 'NAD',
            'segment_group': 'SG12',
            'sub_hierarchy': list([
            ]),
          }),
        ]),
      }),
    ]),
  })
# ---
